### PR TITLE
Sales

### PR DESCRIPTION
**Pull request by cghislai**
- Date Serialization adapter
- Search: closed predicate + pagination + sort columns
- calc ItemSale price on save
- calc Sale price on get/saarch
- remove sale item : delete it
- cancel open sale : delete it
- updated API

### CODE REVIEW

File: comptoir-api-domain/src/main/java/be/valuya/comptoir/api/utils/DateFormatter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ *
+ * @author cghislai
+ */
+public class DateFormatter extends XmlAdapter<String, ZonedDateTime>{
+
+    @Override
+    public ZonedDateTime unmarshal(String string) throws Exception {
+        return ZonedDateTime.parse(string, DateTimeFormatter.ISO_OFFSET_DATE_TIME);
```
Comment (cghislai) [inline comptoir-api-domain/src/main/java/be/valuya/comptoir/api/utils/DateFormatter.java]: This is the ISO format supported by ECMA Date.parse

File: comptoir-ejb/src/main/java/be/valuya/comptoir/service/SaleService.java [hunk 2]
```diff
@@ -272,16 +325,56 @@ public Sale saveSale(Sale sale) {
...
+    public void cancelOpenSale(Sale sale) {
+        if (sale.isClosed()) {
+            throw new IllegalArgumentException("The sale is closed");
+        }
+        List<ItemSale> itemSaleList = findSaleItems(sale);
+        for (ItemSale itemSale : itemSaleList) {
+            removeItemSale(itemSale);
+        }
+        Sale managedSale = entityManager.merge(sale);
+        entityManager.remove(managedSale);
```
Comment (cghislai) [inline comptoir-ejb/src/main/java/be/valuya/comptoir/service/SaleService.java]: I though that for open sales we could remove them completely from the database. I might be wrong, I didn't check all linked entitites were correctly removed, especially the accounting-related ones created in the calc* method.

File: comptoir-ejb/src/main/java/be/valuya/comptoir/service/SaleService.java [hunk 3]
```diff
@@ -272,16 +325,56 @@ public Sale saveSale(Sale sale) {
...
+            itemSale.setDateTime(dateTime);
+        }
+        // Update price
+        BigDecimal quantity = itemSale.getQuantity();
+        Price price = itemSale.getPrice();
+        Price itemPrice = itemSale.getItem().getCurrentPrice();
+        if (price == null) {
+            price = new Price();
+        }
+        BigDecimal vatExclusiveTotal = itemPrice.getVatExclusive().multiply(quantity);
```
Comment (cghislai) [inline comptoir-ejb/src/main/java/be/valuya/comptoir/service/SaleService.java]: Update ItemSale price with the current item price and the quantity amount

File: comptoir-ws/src/main/java/be/valuya/comptoir/ws/rest/control/SaleResource.java [hunk 4]
```diff
@@ -78,14 +95,26 @@ public WsSale getSale(@PathParam("id") long id) {
...
+                .map(saleService::calcSale)
                 .map(toWsSaleConverter::convert)
                 .collect(Collectors.toList());
-        
+        response.setHeader(HeadersConfig.LIST_RESULTS_COUNT_HEADER, count.toString());
         return wsSales;
     }
     
+    @DELETE
+    @Path("{id}/state/OPEN")
```
Comment (cghislai) [inline comptoir-ws/src/main/java/be/valuya/comptoir/ws/rest/control/SaleResource.java]: Meybe you wanted to use path parameters for these state urls

**Comment by ymajoros** [comment]
While merging this pull request, I didn't see that you called calc. Meanwhile I've been working on it too. I mainly removed accounting entries creation, as this should only happen when a sale is validated... Let's see how it merges.

**Comment by cghislai** [comment]
I only called calc from the ws resource on sale get/search methods. it
worked out of the box, except that IIRC it expected only 1 VAT account but
2 were created on company creation.

**Comment by ymajoros** [comment]
I pushed a change to calc sale on Sale / ItemSale update, as those are the only times where it should be calculated.

This is business-related stuff, so shouldn't be called from resource. Resource calls save, which does stuff transparently. Frontend should load sale after Sale / ItemSale update anyway.

Please note that all of this is untested, I just pushed because we were going in somewhat different directions, so we can start again from a common base.

**Pull request merged**